src/main/java/me/karakelley/tictactoe/HardBehavior.java

```diff
@@ -0,0 +1,65 @@
+package me.karakelley.tictactoe;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class HardBehavior implements AIBehavior{
+  private String opponentMarker;
+  private String choice = null;
+
+  public HardBehavior(String opponentMarker) {
+    this.opponentMarker = opponentMarker;
+  }
+
+  public String move(BoardState boardState) {
+    if (blockNeeded(boardState)) {
+      return choice;
+    } else {
+      return availableCells(boardState).get(randomCell(boardState)).toString();
+    }
+  }
+
+  private void checkSection(String[][] section) {
+    for (String[] aSection : section) {
+      if (checkForOpponentMarker(aSection).length == aSection.length - 1) {
+        findBlankCell(aSection);
+      }
+    }
+  }
+
+  private void findBlankCell(String[] section) {
+    List<String> findBlanks = checkForBlankCell(section);
+    if (findBlanks.size() > 0) {
+      choice = Collections.singletonList(findBlanks.get(0)).get(0);
+    }
+  }
+
+  private List<String> checkForBlankCell(String[] section) {
+    return Arrays.stream(section)
+            .filter(space -> space.matches("[0-9]+"))
+            .collect(Collectors.toList());
+  }
+
+  private Object[] checkForOpponentMarker(String[] section) {
+    return Arrays.stream(section)
+            .filter(marker -> marker.equals(opponentMarker))
+            .toArray();
+  }
+
+  private boolean blockNeeded(BoardState boardState) {
+    checkSection(boardState.rows());
+    checkSection(boardState.columns());
+    checkSection(boardState.diagonals());
```

Comment:
this seems like it keeps checking sections even if it already found a blocking move. any reason you're not returning when you found a blocking move?